### PR TITLE
Use Transcript object in sync and async.

### PR DESCRIPTION
**Pull request by daspecster**
In the spirit of smaller PRs. I started switching sync to GAPIC again and I realized that this is required first.

### CODE REVIEW

File: speech/google/cloud/speech/client.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from google.cloud.speech.encoding import Encoding
 from google.cloud.speech.operation import Operation
 from google.cloud.speech.sample import Sample
+from google.cloud.speech.transcript import Transcript
```
Comment (dhermes) [inline speech/google/cloud/speech/client.py]: Worth putting `Transcript` in `__init__.py` as well?

File: speech/google/cloud/speech/client.py [hunk 2]
```diff
@@ -195,7 +196,9 @@ def sync_recognize(self, sample, language_code=None,
             method='POST', path='speech:syncrecognize', data=data)
 
         if len(api_response['results']) == 1:
-            return api_response['results'][0]['alternatives']
+            return [Transcript.from_api_repr(alternative)
+                    for alternative
+                    in api_response['results'][0]['alternatives']]
```
Comment (dhermes) [inline speech/google/cloud/speech/client.py]: While you're here, can you make a separate variable

```python
result = api_response['results'][0]
```

which "confirms" the `len(api_response['results']) == 1` check and then in your list comprehension here you'd just say `result['alternatives']`.

File: speech/google/cloud/speech/transcript.py [hunk 3]
```diff
@@ -16,14 +16,30 @@
...
-        self._confidence = result.get('confidence')
+    def __init__(self, transcript, confidence):
+        self._transcript = transcript
+        self._confidence = confidence
+
+    @classmethod
+    def from_api_repr(cls, transcript):
+        """Factory: construct ``Transcript`` from JSON response.
+
+        :type transcript: :class:`~SpeechRecognitionAlternative`
```
Comment (dhermes) [inline speech/google/cloud/speech/transcript.py]: The `~` is unneccesary if there is no path to truncate. Also, isn't this a `dict`?

File: speech/google/cloud/speech/transcript.py [hunk 4]
```diff
@@ -16,14 +16,30 @@
...
+        :param transcript: Instance of ``SpeechRecognitionAlternative``
+                           from protobuf.
+
+        :rtype: :class:`~Transcript`
+        :returns: Instance of ``Transcript``.
+        """
+        return cls(transcript['transcript'], transcript['confidence'])
```
Comment (dhermes) [inline speech/google/cloud/speech/transcript.py]: If it is a protobuf object why use dict syntax / `__getitem__` here?

File: speech/unit_tests/test_client.py [hunk 5]
```diff
@@ -109,12 +111,16 @@ def test_sync_recognize_content_with_optional_parameters(self):
         self.assertEqual(req['method'], 'POST')
         self.assertEqual(req['path'], 'speech:syncrecognize')
 
-        expected = SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives']
-        self.assertEqual(response, expected)
+        alternative = SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives'][0]
+        expected = [Transcript.from_api_repr(alternative)]
+        self.assertIsInstance(response[0], Transcript)
```
Comment (dhermes) [inline speech/unit_tests/test_client.py]: Add an assertion for the length of the response.

File: speech/unit_tests/test_client.py [hunk 6]
```diff
@@ -109,12 +111,16 @@ def test_sync_recognize_content_with_optional_parameters(self):
...
+        self.assertEqual(response[0].transcript, expected[0].transcript)
+        self.assertEqual(response[0].confidence, expected[0].confidence)
```
Comment (dhermes) [inline speech/unit_tests/test_client.py]: What is the point of making `expected` if you just need to refer to `expected[0]`?

File: speech/unit_tests/test_client.py [hunk 7]
```diff
@@ -144,8 +150,11 @@ def test_sync_recognize_source_uri_without_optional_parameters(self):
         self.assertEqual(req['method'], 'POST')
         self.assertEqual(req['path'], 'speech:syncrecognize')
 
-        expected = SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives']
-        self.assertEqual(response, expected)
+        expected = [Transcript.from_api_repr(
+            SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives'][0])]
+        self.assertIsInstance(response[0], Transcript)
+        self.assertEqual(response[0].transcript, expected[0].transcript)
+        self.assertEqual(response[0].confidence, expected[0].confidence)
```
Comment (dhermes) [inline speech/unit_tests/test_client.py]: Ditto here.

File: speech/google/cloud/speech/client.py (hunk 2 above)
Comment (daspecster) [inline speech/google/cloud/speech/client.py]: How about...
```python
        results = api_response['results']
        if len(results) == 1:
            result = results[0]['alternatives']
            return [Transcript.from_api_repr(alternative)
                    for alternative in result]
```

File: speech/google/cloud/speech/transcript.py (hunk 4 above)
Comment (daspecster) [inline speech/google/cloud/speech/transcript.py]: Copy pasta :(   It's not protobuf in `from_api_repr`.

File: speech/google/cloud/speech/client.py (hunk 2 above)
Comment (dhermes) [inline speech/google/cloud/speech/client.py]: The point of my suggestion was to have a line of code ending in `[0]` to emphasize that we have exactly one.

File: speech/unit_tests/test_client.py [hunk 8]
```diff
@@ -112,10 +112,11 @@ def test_sync_recognize_content_with_optional_parameters(self):
         self.assertEqual(req['path'], 'speech:syncrecognize')
 
         alternative = SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives'][0]
-        expected = [Transcript.from_api_repr(alternative)]
+        expected = Transcript.from_api_repr(alternative)
+        self.assertEqual(len(response, 1))
```
Comment (dhermes) [inline speech/unit_tests/test_client.py]: Your parentheses are out of order: `self.assertEqual(len(response), 1)`

File: speech/google/cloud/speech/client.py [hunk 9]
```diff
@@ -195,10 +195,11 @@ def sync_recognize(self, sample, language_code=None,
         api_response = self.connection.api_request(
             method='POST', path='speech:syncrecognize', data=data)
 
-        if len(api_response['results']) == 1:
+        results = api_response['results']
+        if len(results) == 1:
+            result = results[0]['alternatives']
```
Comment (dhermes) [inline speech/google/cloud/speech/client.py]: Change it to `results[0]` if you like. This is "clear enough" but doesn't emphasize the "first item" part as I'd hoped. Not a dealbreaker though.

File: speech/unit_tests/test_client.py [hunk 10]
```diff
@@ -150,11 +151,11 @@ def test_sync_recognize_source_uri_without_optional_parameters(self):
         self.assertEqual(req['method'], 'POST')
         self.assertEqual(req['path'], 'speech:syncrecognize')
 
-        expected = [Transcript.from_api_repr(
-            SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives'][0])]
+        expected = Transcript.from_api_repr(
+            SYNC_RECOGNIZE_RESPONSE['results'][0]['alternatives'][0])
```
Comment (dhermes) [inline speech/unit_tests/test_client.py]: You also want a length check here.

File: speech/google/cloud/speech/client.py (hunk 2 above)
Comment (daspecster) [inline speech/google/cloud/speech/client.py]: I didn't catch that, I'm pushing your original suggestion.

**Comment by daspecster** [comment]
Ok thanks! I changed the response check via your suggestion, added the length check and fixed my bug.

I also just squashed the commits.

**Comment by daspecster** [comment]
Travis is green.  Merging.

**Pull request merged**